pageserver/src/walredo.rs

```diff
@@ -228,34 +229,30 @@ impl PostgresRedoManager {
         key: Key,
         lsn: Lsn,
         base_img: Option<Bytes>,
-        records: &[(Lsn, NeonWalRecord)],
+        records: &std::sync::Arc<[(Lsn, NeonWalRecord)]>,
+        records_range: SliceRange,
         wal_redo_timeout: Duration,
         pg_version: u32,
     ) -> Result<Bytes, WalRedoError> {
         let (rel, blknum) = key_to_rel_block(key).or(Err(WalRedoError::InvalidRecord))?;
 
         let start_time = Instant::now();
 
-        let mut process_guard = self.process.lock().unwrap();
-        let lock_time = Instant::now();
-
-        // launch the WAL redo process on first use
-        if process_guard.is_none() {
-            let p = PostgresRedoProcess::launch(self.conf, self.tenant_id, pg_version)?;
-            *process_guard = Some(p);
-        }
-        let process = process_guard.as_mut().unwrap();
-
-        WAL_REDO_WAIT_TIME.observe(lock_time.duration_since(start_time).as_secs_f64());
-
         // Relational WAL records are applied using wal-redo-postgres
         let buf_tag = BufferTag { rel, blknum };
-        let result = process
-            .apply_wal_records(buf_tag, base_img, records, wal_redo_timeout)
-            .map_err(WalRedoError::IoError);
 
-        let end_time = Instant::now();
-        let duration = end_time.duration_since(lock_time);
+        let result = self
```

Comment:
Add comment that this `block_on` requires that we are:
- entered into tokio context
- outside of async context

Basically within `spawn_blocking` which #2873 provides.

Reply:
Alternative is to use tokio::task::block_in_place somewhere higher up which should have less of a perf penalty, at the cost runtime of starvation hazard. This is being tested now, resolving and moving these as PR tasks.